src/components/Tooltip/Tooltip.tsx

```diff
@@ -0,0 +1,347 @@
+import React, {
+  createElement,
+  ForwardRefExoticComponent,
+  ReactElement,
+  ReactNode,
+  RefObject,
+  useEffect,
+  useRef,
+  useState,
+} from 'react'
+import classnames from 'classnames'
+import { isElementInViewport } from './utils'
+
+type TooltipProps<T> = {
+  label: string
+  position?: 'top' | 'bottom' | 'left' | 'right' | undefined
+  className?: string
+  dataClasses?: string
+  children: ReactNode
+} & T
+
+interface WithCustomTooltipProps<T> {
+  asCustom: ForwardRefExoticComponent<T>
+}
+
+export type DefaultTooltipProps = TooltipProps<JSX.IntrinsicElements['button']>
+
+export type CustomTooltipProps<T> = TooltipProps<T> & WithCustomTooltipProps<T>
+
+export function isCustomProps<T>(
+  props: DefaultTooltipProps | CustomTooltipProps<T>
+): props is CustomTooltipProps<T> {
+  return 'asCustom' in props
+}
+
+const tooltipID = `tooltip-${Math.floor(Math.random() * 900000) + 100000}`
```

Comment:
Because this `tooltipID` is defined outside of the `Tooltip` component, it's the same value for all instances of a Tooltip used within a single app. This should be defined inside of the `Tooltip` component in a `useRef` container:
```
const tooltipID = useRef(`tooltip-${Math.floor(Math.random() * 900000) + 100000}`)

// to access:
tooltipID.current
```